File: compleat/c++/with_Hidden/with_h/generative_model_hidden.cpp
// Generative model of potts. 
#include <cmath>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <iomanip>
#include <random>
#include <numeric>
using namespace std;

int q=3, L=16;
int p=L;
double Xi0=0.1, h0 = 1.0;

/*****************************/
vector<int> X(L,0);
vector<double> H(p,0);
vector < vector<double> > Xi( L*p, vector<double>(q, Xi0) );  
vector < vector<double> > h( L, vector<double>(q,0) );  
//vector < vector<double> > Psi( L*p, vector<double>(q,0) );  
//vector < vector<double> > Psi_h( L, vector<double>(q,0.0) );  
/*****************************/
int T_equil = 1e3;
int T_interval = 1e2;
//int N_sampling = 1e5;
int N_sampling = 1e4;
/*****************************/
int K = 1;
/*****************************/

inline double std_rand()
{
    return rand() / (RAND_MAX + 1.0);
}

void X_set(){
	for(int i=0;i<L;++i){
		X[i] = int(q*std_rand()) ;
	}
}

void set_h_Xi(){
	for(int i=0;i<L;++i){
		for(int a=0;a<q;++a){
			h[i][a] = h0 * std_rand();
			for(int m=0;m<p;++m){
				Xi[i*p+m][a] = Xi0 * std_rand();
			
	}}}
}

bool Xi_init(){
	bool flag=true;
	string fname_in = "Xi_model_estimator_CD_L"+to_string(L)+"_Hidden_c++.dat";
  	ifstream file(fname_in);
  	string line;
  	int i=0,m=0,a=0; 
	if (file.is_open()){
    	while ( getline (file,line,'\n'))
    	{
		Xi[i*p+m][a] = stod(line);
		m += 1;
		if(m==p){//NOTE: m moves faster than a.
			a+=1; m=0;	
			if(a==q){
				i+=1; a=0;
			}
		}		
	}	
    	file.close();
	flag = true;
	}
  	else{ cout << "Unable to open file" << endl;  
  	cout << fname_in << endl;
    	flag = false;
  	}
	return flag;
}

bool h_init(){
	bool flag=true;
	string fname_in = "h_model_estimator_CD_L"+to_string(L)+"_Hidden_c++.dat";
  	ifstream file(fname_in);
  	string line;
  	int count = 0; int i=0,a=0,var; // this initialize is necessary.
		
  	if (file.is_open()){
    	while ( getline (file,line,'\n'))
    	{
		h[i][a] = stod(line);
		a += 1;
		if(a==q){
			i+=1; a=0;
		}	
	}
    	file.close();
	flag = true;
	}
  	else{ cout << "Unable to open file" << endl;  
  	cout << fname_in << endl;
    	flag = false;
  	}
	
	return flag;
}

void get_normaldist(vector<double> mean, double std){
	random_device rd;
	mt19937 gen(rd());
	for(int m=0;m<p;++m){
		normal_distribution<> d(mean[m],std);
		H[m] = d(gen);
		//cout << m << " " << mean[m] << " " << H[m] << endl;	
	}		
}

int my_sampling_gibbsdist(vector<double> prob){
	random_device rd;
	mt19937 gen(rd());
	using dist_type = discrete_distribution<>;
	discrete_distribution<> d(prob.begin(), prob.end());
	int resolute = d(gen);
	return resolute; 
}

double E_local_hidden(int i,int a){
	double E_i = 0.0;
	for(int m=0;m<p;++m){
		E_i += - Xi[i*p+m][a]*H[m];
	}
	E_i += -h[i][a];
	return E_i;	
}

void sampling_hidden(){
	int a;
	vector<double> x_0(p,0);
	for(int m=0;m<p;++m){
		for(int i=0;i<L;++i){
			a = int(X[i]);
			x_0[m] += Xi[i*p+m][a];
		}
		x_0[m] /= double(L);
	}
	get_normaldist(x_0, 1.0/L);// mean=x_0, std=1/L, size=p
}

void sampling_visible(){
	double E_min,pdis_sum;
	vector<double> pdis(q,0);
	vector<double> E_vec(q,0);
	
	for(int i=0;i<L;++i){
	 	for(int a=0;a<q;++a){	E_vec[a] = E_local_hidden(i,a);	}
		E_min = *min_element(E_vec.begin(), E_vec.end());
		pdis_sum = 0.0;
		for(int a=0;a<q;++a){
			pdis[a] = exp( - (E_local_hidden(i,a) - E_min) );	
			pdis_sum += pdis[a];
		}
		//pdis_sum = accumulate(pdis.begin(), pdis.end(), 0);
	 	for(int a=0;a<q;++a){ pdis[a] /= pdis_sum ; }
		X[i] = my_sampling_gibbsdist(pdis);	
	}
}

void outputestimator(){
    	// Coupling Constant 
	string fname = "Xi_training_L"+to_string(L)+"_Hidden_c++.dat";
	ofstream file(fname);
	string fname_h = "h_training_L"+to_string(L)+"_Hidden_c++.dat";
	ofstream file_h(fname_h);
	
	for(int i=0;i<L;i++){
	for(int a=0;a<q;++a){
		file_h << h[i][a] << endl; 
		for(int m=0;m<p;++m){
			file << Xi[i*p+m][a] << endl; 
		}}}
	file.close(); 
	file_h.close();
}

int main(){
	h_init();
	Xi_init();	
	
	/* 
	//set_h_Xi();
	//outputestimator();	
	*/
	
	//string fname = "genmodel_data_ML_L"+to_string(L)+"_c++.dat" ;
    	string fname  = "test_training_data_L"+to_string(L)+"_Hidden_c++.dat";
	ofstream file(fname);
	int t;
	t = 0;
	//X_set();
	while(t<T_equil){
		sampling_hidden();
		sampling_visible();
		t += 1;	
	}	
	
	t = 0;
	while(t<N_sampling*T_interval){
		sampling_hidden();
		sampling_visible();
		if(t%T_interval == 0){
			for(int i=0;i<L;++i){
				file << X[i] << " ";	
			}
			file << endl;
		}
		t+=1;
	}
	file.close();
}

File: compleat/c++/with_Hidden/with_h/generative_model_hidden.py
# -*- coding: utf-8 -*- 
import numpy as np
import matplotlib.pyplot as plt

q, L = 3, 4 
p = L
X = np.random.choice(q,L) # Chose L-variables from q-state .
Xi0, h0 =  0.3, 0.3 
h = np.zeros((L,q))
Xi = np.ones((L*p,q)) * Xi0

N_sampling = int(1e4) 
T_equil = int(1e3)
t_relax = 30

Xi_data = np.ones((L*p,q)) * Xi0
H = np.zeros(p) # Chose L-variables from q-state .
def Xi_set():
    global Xi_data
    fname_in = "Xi_model_estimator_CD_L"+str(L)+"_Hidden.dat"
    #fname_in = "Xi_model_estimator_CD_L"+str(L)+"_Hidden_c++.dat"
    fin = open(fname_in, "r")
     
    a=0; m=0; i=0; 
    for line in fin:
        line = line.replace('\n','') # do not need to use 
        item = line.split(' ')
        
        Xi_data[i*p+m][a]= float(item[0])
        a += 1
        if(a==q):
            a = 0 
            m += 1
            if(m==p):
                m = 0
                i += 1
    fin.close()

def h_init():
    global  h
    #fname_in = "h_model_estimator_L"+str(L)+".dat"
    fname_in = "h_model_estimator_CD_L"+str(L)+"_Hidden.dat"
    #fname_in = "h_model_estimator_CD_L"+str(L)+"_Hidden_c++.dat"
    fin = open(fname_in, "r")
    a=0; i=0 
    for line in fin:
        line = line.replace('\n','')
        item = line.split(' ')
        h[i][a] = float(item[0]) 
        a += 1
        if(a==q):
            a=0
            i+=1
    fin.close()

def E_local_hidden(i,a):
    global Xi_data, h
    E_i = .0
    for m in range(p): # NOTE: J[i*L+j][xx] = 0
        E_i += - Xi_data[i*p+m][a]*H[m]
    E_i += -h[i][a] 
    return E_i

def sampling_hidden():
    global X, H, Xi_data
    x_0 = np.zeros(p)
    for m in range(p):
        x_0[m] = 0.0
        for i in range(L):
            a = int(X[i] ) 
            x_0[m] += Xi_data[i*p+m][a]
        x_0[m] /= L
    H = np.random.normal(x_0,1.0/L,p)    

def sampling_visible():
    global X
    pi = np.zeros(q)
    X = [] 
    for i in range(L):
        
        E_vec = np.zeros(q)
        for a in range(q):
            E_vec[a] =  E_local_hidden(i,a)
        E_min = np.min(E_vec)
        
        pdis = np.zeros(q) 
        for a in range(q):
            pdis[a] = np.exp(- ( E_local_hidden(i,a) - E_min ))

        pdis_sum = np.sum(pdis)
        pdis = pdis / pdis_sum
        # Does this p give effect on p(= #pattern)?
        X.append( int(np.random.choice(np.arange(q),p=pdis)) ) 

if __name__ == "__main__":
    h_init()
    Xi_set()
    fname_Data  = "test_training_data_L"+str(L)+"_Hidden.dat"
    #fname_Data  = "test_training_data_L"+str(L)+"_Hidden_c++.dat"
    fout_Data = open(fname_Data,"w")
      
    t = 0
    while(t<T_equil):
        sampling_hidden()
        sampling_visible()
        t += 1

    #X = np.random.choice(q,L)
    t =  0
    while(t<N_sampling*t_relax):
        sampling_hidden()
        sampling_visible()
        if(t%t_relax == 0):
            for i in range(L):
                fout_Data.write(str(X[i])+" ")
            fout_Data.write("\n")
        t += 1
    fout_Data.close() 

